packages/database/src/cachers/EntityHierarchyCacher.js

```diff
@@ -40,105 +34,115 @@ export class EntityHierarchyCacher {
+    const childIdsToAncestorIds = await this.fetchChildIdsToAncestorIds(
       hierarchyId,
-      parentIds,
+      parentIdsToAncestorIds,
+      useEntityRelationLinks,
     );
 
-    const childIdsToAncestorIds = {};
-    for (const parentId of parentIds) {
-      const ancestorIds = parentIdsToAncestorIds[parentId];
-      const childIds = nextGenerationIsCanonical
-        ? await this.getNextGenerationCanonically(parentId)
-        : await this.getNextGenerationViaEntityRelation(hierarchyId, parentId);
-
-      // if childIds is empty, we've made it to the leaf nodes
-      if (childIds.length === 0) {
-        continue;
-      }
-
-      // cache this generation
-      const parentAndAncestorIds = [parentId, ...ancestorIds];
-      await this.cacheGeneration(hierarchyId, childIds, parentAndAncestorIds);
-
-      // add this generation to the object for caching in the next recursion
-      childIds.forEach(childId => {
-        childIdsToAncestorIds[childId] = parentAndAncestorIds;
-      });
+    const childrenAlreadyCached = await this.checkChildrenAlreadyCached(
+      hierarchyId,
+      parentIds,
+      childCount,
+    );
+    if (!childrenAlreadyCached) {
+      await this.cacheGeneration(hierarchyId, childIdsToAncestorIds);
     }
 
-    // keep recursing through the hierarchy
-    await this.fetchAndCacheDescendants(hierarchyId, childIdsToAncestorIds);
+    // if there is another generation, keep recursing through the hierarchy
+    await this.fetchAndCacheDescendants(
+      hierarchyId,
+      childIdsToAncestorIds,
+      entityRelationChildCount === 0,
+    );
   }
 
-  async checkShouldFetchAndCache(hierarchyId, parentIds) {
-    if (parentIds.length === 0) {
-      return false; // base case of recursion, we must have reached the leaf nodes
-    }
-
-    // check whether this generation/hierarchy combo has already been cached to avoid doing it again
-    // on startup, or when two projects share a hierarchy (at time of writing none do, but db schema
-    // makes it possible)
-    const numberAlreadyCached = await this.models.ancestorDescendantRelation.count({
+  async checkChildrenAlreadyCached(hierarchyId, parentIds, childCount) {
+    const numberChildrenCached = await this.models.ancestorDescendantRelation.count({
       entity_hierarchy_id: hierarchyId,
       ancestor_id: parentIds,
       generational_distance: 1,
     });
-    return numberAlreadyCached !== parentIds.length;
+    return numberChildrenCached === childCount;
+  }
+
+  async fetchChildIdsToAncestorIds(hierarchyId, parentIdsToAncestorIds, useEntityRelationLinks) {
+    const parentIds = Object.keys(parentIdsToAncestorIds);
+    const relations = useEntityRelationLinks
+      ? await this.getRelationsViaEntityRelation(hierarchyId, parentIds)
+      : await this.getRelationsCanonically(parentIds);
+    const childIdToParentId = reduceToDictionary(relations, 'child_id', 'parent_id');
+    const childIdsToAncestorIds = Object.fromEntries(
```

Comment:
I don't know if you agree with this change? I believe it's a good indication of the `childId => ancestorIds` relationship in this object.

By the way, there should really be a universal naming guide for variables, similar to code style guides 🤓 Lots of the patterns and structures are often repeated!

```suggestion
    const childIdToAncestorIds = Object.fromEntries(
```

Reply:
Yeah I tend to agree with you on the singular for childId, it matches the naming conventions we've used elsewhere. 

Perhaps a good dev chat item - naming conventions! I've noticed some of our common tupaia conventions may not be natural to devs coming from other companies. A good example is that others tend to use 
```
async function getXByY(y) { 
  const x = await db.fetchX();
  return x.filter(x => x.y === y); 
}
```
whereas we would expect the same name to do something like
```
async function getXByY() { 
  const x = await db.fetchX();
  return groupBy(x, y); 
}
```

Reply:
Agree! Sounds like a valuable documentation, it's not something that changes every day